remove timeout from `send_to_closest` + send to `self`

RolandSherwin opened the pull request:
fixes #193 
- fix(network): get our `PeerId` from Network
    - use `network.peer_id` instead of storing it around in `PendingGetClosest`
- fix(network): remove timeout from `send_to_closest`
    - The request_response Behaviour contains an inbuilt timeout. Hence, remove our custom timeout implementation.
- feat(network): set timeout through the `RequestResponse` behaviour
- fix(network): route `Request` and `Response` to self 
	- While using the `RequestResponse` behaviour, we get a `OutboundFailure::DialFailure`
if we try to send a request to `self`
	- So if `self` is the recipient of the `Request`, then route the request
  directly to `self` without using the `RequestResponse` behaviour.
	- This request, then follows the normal flow without having any custom
  branch on the upper layers. The produced `Response` is also routed
  back to `self`
- fix(test): initialize logging once for unit tests
	- else the test panics as we initialize multiple `Subscriber` instances, at multiple sub-processes when we call `cargo test`

RolandSherwin committed:
fix(network): get our `PeerId` from Network

RolandSherwin committed:
fix(network): remove timeout from `send_to_closest`

- The request_response Behaviour contains an inbuilt timeout. Hence
  remove our custom timeout implementation.

RolandSherwin committed:
feat(network): set timeout through the `RequestResponse` behaviour

oetyng commented on safenode/src/network/mod.rs:
```diff
@@ -145,26 +144,31 @@ impl SwarmDriver {
+        let _ = req_res_config.set_request_timeout(REQUEST_TIMEOUT);
+        let _ = req_res_config.set_connection_keep_alive(CONNECTION_KEEP_ALIVE_TIMEOUT);
+
+        let request_response = request_response::Behaviour::new(
```
Shouldn't this variable be called `behaviour`?
`request_response` is a very fuzzy name. It is a `Behaviour` instance, right?

RolandSherwin replied:
I think we have a couple of other structs that implement the `Behaviour` trait, and they're just called by their crate/struct name, e.g. `kademlia`, `mdns`. So I think `request_response` fits well in this context? 

The variable names are self-explanatory inside this function, but adding `_behaviour` to them is also an option.

oetyng replied:
Yeah, will mull this over, later.. :)

oetyng commented on safenode/src/network/mod.rs:
```diff
@@ -296,50 +300,31 @@ impl Network {
-            "Got the {} closest_peers to the given XorName-{xor_name}, nodes: {closest_peers:?}",
-            closest_peers.len()
+    /// Send `Request` to the closest peers. `Self` might be present among the recipients.
+    pub async fn node_send_to_closest(&self, request: &Request) -> Result<Vec<Result<Response>>> {
```
It would be nice to have traits for shared objects such as Network, that don't expose apis that are not intended for certain layers of the application, such as client or node.

oetyng replied:
So when as a client using the network it won't list all apis that shouldn't be used by a client, only those that are supposed to be used by it.

RolandSherwin replied:
Yeah, that would be great and would lead to some concise function names. Do you mind if I create an issue for it and solve it on a follow-up PR?

oetyng replied:
Please do! Thanks!

RolandSherwin replied:
Raise issue https://github.com/maidsafe/safe_network/issues/203!

oetyng commented on safenode/src/network/mod.rs:
```diff
@@ -296,50 +300,31 @@ impl Network {
+            request.dst().name()
         );
+        // todo: if `self` is present among the closest peers, the request should be routed to self?
+        let closest_peers = self.node_get_closest_peers(*request.dst().name()).await?;
```
I wonder if we quite know what happens with messages sent to ourselves?

Can we setup a test for it?

Just to confirm that we can send any number of msgs to ourselves and it will just flow through without any deadlock or funny loops or such.

RolandSherwin replied:
> Just to confirm that we can send any number of msgs to ourselves and it will just flow through without any deadlock or funny loops or such.

The msg to `self` indeed causes an issue, as confirmed by a test. I have updated the PR with a fix for it.

RolandSherwin committed:
fix(network): route `Request` and `Response` to self
- While using the `RequestResponse` behaviour, we get a `OutboundFailure::DialFailure`
if we try to send a request to `self`
- So if `self` is the recipient of the `Request`, then route the request
  directly to `self` without using the `RequestResponse` behaviour.
- This request, then follows the normal flow without having any custom
  branch on the upper layers. The produced `Response` is also routed
  back to `self`

RolandSherwin committed:
fix(test): initialize logging once for unit tests

oetyng reviewed (approved):
Nice one!!

RolandSherwin committed:
fix(test): keep the event_rx task running

oetyng reviewed (changes requested):
This PR is unfortunately not solving the node many to many msg issue. Or perhaps it is but causing other issues instead. Hard to say, but network is even less stable than before, unfortunately :(

oetyng committed:
test: fix msg_to_self_should_not_error_out

- Some retrying makes it pass every time.

oetyng committed:
chore: minor style fixes

oetyng reviewed (approved):
Ok, test changes have been made, and the test passes. Unless any other tests fail, let's get this in.

And I'll deal with the effects for transfer when we cross that bridge (right up after this actually..).
